parsers/NTESMO.py

```diff
@@ -177,93 +183,151 @@ def parse_consumption(
-        )
-    raw_production_mix["Period Start"] = raw_production_mix[
+    raw_production_mix_df["Period Start"] = raw_production_mix_df[
         "Period Start"
     ].dt.tz_localize("Australia/Darwin")
-    for _, production in raw_production_mix.iterrows():
-        data_point = {
-            "zoneKey": "AU-NT",
-            "datetime": production["Period Start"].to_pydatetime(),
+
+    # some older data may miss columns for newer plants
+    raw_production_mix_df = raw_production_mix_df.dropna(axis=1, how="all")
+
+    generation_units = set(raw_production_mix_df.columns)
+    generation_units.remove("Period Start")
+    generation_units.remove("Period End")
+    unknown_units = generation_units.difference(PLANT_MAPPING.keys())
+    logger.warning(
+        f"New generator(s) {unknown_units} detected in AU-NT, please update the mapping of generators."
+    )
+
+    production_breakdown_list = []
+    for row in raw_production_mix_df.itertuples():
+        production_mix = collections.defaultdict(float)
+        for generator_key in generation_units:
+            generator_production = getattr(row, generator_key)
+
+            # Some decommissioned plants have negative production values.
+            if generator_production < 0:
+                continue
+
+            fuel_type = (
+                "unknown"
+                if generator_key in unknown_units
+                else PLANT_MAPPING[generator_key]["fuel_type"]
+            )
+            production_mix[fuel_type] += generator_production
+
+        production_breakdown = {
+            "zoneKey": ZONE_KEY,
+            "datetime": row[1].to_pydatetime(),
             "source": "ntesmo.com.au",
-            "production": {},
+            "production": dict(production_mix),
             "storage": {},
         }
-        for generator_key, generator in PLANT_MAPPING.items():
-            if generator_key not in production:
-                raise ParserException(
-                    "NTESMO.py",
-                    f"Missing generator {generator_key} detected in AU-NT, please update the mapping of generators.",
-                )
-            # Some decomissioned plants have negative production values.
-            if production[generator_key] >= 0:
-                if generator["fuel_type"] in data_point["production"]:
-                    data_point["production"][generator["fuel_type"]] += production[
-                        generator_key
-                    ]
-                else:
-                    data_point["production"][generator["fuel_type"]] = production[
-                        generator_key
-                    ]
-        production_mix.append(data_point)
-    return production_mix
+        production_breakdown_list.append(production_breakdown)
+
+    return production_breakdown_list
 
 
 @refetch_frequency(timedelta(days=1))
-@retry_policy(retry_policy=retry_strategy)
```

Comment:
Why where these moved/removed?

I'm not sure it will work as expected if they are as the backend only really look at the top function declarations (the ones that are imported for the parser to run).